config: move enable_save_crystals to config; add override

walkawayy opened the pull request:
#### Checklist

- [X] I have read the [coding conventions](https://github.com/rr-/Tomb1Main/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description
Moved the enable_save_crystals option from the gameflow to the config tool. Added a gameflow option to force a mode for custom levels. Resolves #963. In a future PR, I will rename `disable_game_modes` to follow this naming convent (`enable_...`) and do an `&` with the gameflow and config option.

walkawayy committed:
config: move enable_save_crystals option to config

Resolves #963.

rr- reviewed (commented):
This is a breaking change; level builders now can't turn this feature off. Furthermore it makes little sense to have crystals mode in levels that do not support it.

IMO it should be present in both config and gameflow so that the level can choose if they want to support this feature and the user can choose if they can go with that.

rr- reviewed (commented):
We should avoid manipulating other objects (here: the config) inside of a function that's about loading the gameflow. It's acceptable to do it after ithe internals finish loading in the entrypoint function where it's much more visible. 

There's no way to force save crystals which is guaranteed to be requested by the builders. I think we should employ the tristate idiom here:

| User config | Gameflow settings | Outcome |
| --- | --- | --- |
| `false` | `null` or missing | Crystals off |
| `false` | `false` | Crystals off |
| `false` | `true` | Crystals on |
| `true` | `null` or missing | Crystals on |
| `true` | `false` | Crystals off |
| `true` | `true` | Crystals on |

walkawayy commented:
I can move the config manipulation to another function. Do you have a location in mind or should I propose a spot?

Also this gameflow setting does indeed force crystals if true. So it behaves like your table.

rr- commented:
> Do you have a location in mind or should I propose a spot?

For this particular case, the entrypoint function, i.e. the outermost public-facing function, is `GameFlow_LoadFromFile`. Given its involvement in global string manipulation, it appears to be a suitable choice for manipulating `g_Config`.

>Also this gameflow setting does indeed force crystals if true. So it behaves like your table.

I have some concerns regarding this – the table I proposed distinguishes between a `false` value and a `null` or missing value. Moreover, it doesn't seem to directly offer overriding specific states. So, given the current code,

```c
    g_GameFlow.enable_save_crystals_option =
￼        json_object_get_bool(obj, "enable_save_crystals_option", false);
￼    g_Config.enable_save_crystals &= g_GameFlow.enable_save_crystals_option;
```

here's how I understand the state table looks:

| User setting | Gameflow setting  | Current outcome | Desired outcome |
| ---          | ---               | ---             | ---             |
| `false`      | `null` or missing | `false`         | `false`         |
| `false`      | `false`           | `false`         | `false`         |
| `false`      | `true`            | `false`         | `true`          |
| `true`       | `null` or missing | `false`         | `true`          |
| `true`       | `false`           | `false`         | `false`         |
| `true`       | `true`            | `true`          | `true`          |

This is particularly problematic that for custom levels that do not supply this setting, it is not possible to play with save crystals enabled. Could you please confirm if that's the case?

walkawayy commented:
Oh I misinterpreted. I see what you would like and updated the PR to reflect that.

I moved the check to `GameFlow_LoadFromFile`. I made it so the gameflow option `force_enable_save_crystals` will force save crystals when set to true. In any other case of missing or false, it leaves it up to the user's preference. This makes the most sense to me IMO. The builder can force crystals if needed, otherwise the user chooses. How it behaves in my newly pushed PR:

| User setting | Gameflow setting  | Current outcome | Desired outcome |
| ---           | ---                      | ---        | ---        |
| `false`      | `null` or missing | `false`   | `false`   |
| `false`      | `false`                 | `false`   | `false`   |
| `false`      | `true`                  | `true`    | `true`    |
| `true`       | `null` or missing | `true`    | `true`    |
| `true`       | `false`                 | `true`    | `true`   |
| `true`       | `true`                  | `true`    | `true`    |

walkawayy committed:
rename option, move gameflow and config check, tweak behavior

rr- reviewed (approved):
LGTM
